src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp

```diff
@@ -37,27 +44,51 @@ bool MergeTreePrefetchedReadPool::TaskHolder::operator<(const TaskHolder & other
     return task->priority > other.task->priority; /// Less is better.
 }
 
-MergeTreePrefetchedReadPool::PrefetechedReaders::PrefetechedReaders(
+
+MergeTreePrefetchedReadPool::PrefetchedReaders::~PrefetchedReaders()
+{
+    for (auto & prefetch_future : prefetch_futures)
+        if (prefetch_future.valid())
+            prefetch_future.wait();
+}
+
+MergeTreePrefetchedReadPool::PrefetchedReaders::PrefetchedReaders(
     MergeTreeReadTask::Readers readers_,
     Priority priority_,
     MergeTreePrefetchedReadPool & pool_)
     : is_valid(true)
     , readers(std::move(readers_))
 {
-    prefetch_futures.push_back(pool_.createPrefetchedFuture(readers.main.get(), priority_));
+    try
+    {
+        prefetch_futures.push_back(pool_.createPrefetchedFuture(readers.main.get(), priority_));
+
+        for (const auto & reader : readers.prewhere)
+            prefetch_futures.push_back(pool_.createPrefetchedFuture(reader.get(), priority_));
+
+        fiu_do_on(FailPoints::prefetched_reader_pool_failpoint,
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Failpoint for prefetched reader enabled");
+        });
+    }
+    catch (...) /// in case of memory exceptions we have to wait
+    {
+        for (auto & prefetch_future : prefetch_futures)
```

Comment:
So we create a future with 
```
pool_.createPrefetchedFuture(readers.main.get(), priority_))
```
and then push it to the vector
what if we get OOM when allocating extra space inside vector?
we won't wait for the created future AFAIS

Reply:
I think simple `reserve` is enough to avoid it